crates/graphql/src/schema/dataloaders/batcher.rs

```diff
@@ -0,0 +1,120 @@
+use std::{collections::HashMap, sync::Arc};
+
+use indexer_core::db::query_dsl::methods::LoadQuery;
+
+use super::prelude::*;
+
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum Error {
+    #[error("Failed to connect to the database")]
+    ConnectionFailed,
+    #[error("Failed to query the database")]
+    QueryFailed,
+    #[error("Failed to process a model: {0}")]
+    ModelConvert(#[from] Arc<indexer_core::error::Error>),
+}
+
+impl From<indexer_core::error::Error> for Error {
+    fn from(e: indexer_core::error::Error) -> Self {
+        Self::ModelConvert(Arc::new(e))
+    }
+}
+
+pub type BatchResult<T> = Result<T, Error>;
+pub type BatchMap<K, V> = HashMap<K, BatchResult<V>>;
+pub type Loader<K, V> = dataloader::non_cached::Loader<K, BatchResult<V>, Batcher>;
+
+/// Helper trait for merging results into a [`BatchFn`] value
+pub trait BatchExtend: Default {
+    type Element;
+
+    fn extend(&mut self, element: Self::Element);
+}
+
+impl<T> BatchExtend for Option<T> {
+    type Element = T;
+
+    fn extend(&mut self, element: T) {
+        *self = Some(element);
+    }
+}
+
+impl<T> BatchExtend for Vec<T> {
+    type Element = T;
+
+    fn extend(&mut self, element: T) {
+        self.push(element);
+    }
+}
+
+#[derive(Clone)]
+pub struct Batcher(Arc<Pool>);
+
+impl Batcher {
+    #[must_use]
+    pub fn new(pool: Arc<Pool>) -> Self {
+        Self(pool)
+    }
+
+    pub fn diesel_batch<
+        K: Clone + Eq + std::hash::Hash,
+        KR: Into<K>,
+        V: BatchExtend,
+        M,
+        Q: LoadQuery<Connection, M>,
+        E: Into<indexer_core::error::Error>,
+    >(
```

Comment:
Chatted in the office but like to ship the batcher trait without `diesel_batch` to start. Lets go with some code duplication for now and drop the result until we update data batcher to support it.